### PR TITLE
[v0.11.1] Fix controller setup - add schemas based on CRDs available

### PR DESCRIPTION
**Pull request by israel-hdez**
**What this PR does / why we need it**:

In #233 an enhancement was done to KServe controllers to watch resources based on available CRDs.

A similar change in the setup of the manager was overlooked: it is also needed to add schemas to the manager based on available CRDs rather than only the values in the inferenceservice-config ConfigMap. This would keep both manager setup and controller setup in sync with regards schemas and watches around the CRDs.

This is, basically, a cherry-pick of commti 5180ec56f25186ef643bb909adaf4784fe337bb5 from upstream PR (not merged yet).

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes https://issues.redhat.com/browse/RHOAIENG-4196

**Type of changes**
Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)

**Feature/Issue validation/testing**:

Recommended testing is by using OpenShift.

* Checkout the code from this pull request (use the method you like)
* Deploy OSSM to your cluster by running `./test/scripts/openshift-ci/deploy.ossm.sh` (available in this repo / the checked out code).
* Deploy Serverless to your cluster by running `./test/scripts/openshift-ci/deploy.serverless.sh`
* Create the odh namespace: `oc new-project opendatahub`
* Deploy the buggy KServe image: `oc apply -k 'github.com/red-hat-data-services/kserve/config/overlays/odh?ref=rhoai-2.8'`

Once KServe is deployed, you should be able to look at the logs and it is going to be running fine. Let's do the problematic setup:

* Undeploy Serverless: `oc delete knativeserving -n knative-serving knative-serving`
* Undeploy OSSM: `oc delete smcp -n istio-system basic`
* Patch the KServe ConfigMap: `oc patch cm -n opendatahub inferenceservice-config --type merge -p '{"data": {"deploy": "{\"defaultDeploymentMode\": \"RawDeployment\"}"}}'`
* Restart the kserve controller pod: `oc delete pod -n opendatahub -l control-plane=kserve-controller-manager`

Once it restarts, you should notice `kind must be registered to the Scheme` errors in the logs of the pod and it will enter in a CrashLoop state. Now, deploy the fixed image:

* Build the image by yourself:
  * Declare the IMG envvar to your own container registry; e.g. `export IMG=quay.io/edgarhz/kserve-controller:pr-265`
    * This tag actually exists, in case you prefer to use this pre-built image.
  * Build the image and push it to your registry: `make docker-build docker-push`
* Deploy the fixed image: `oc set image -n opendatahub deploy/kserve-controller-manager manager=$IMG`

Once the KServe pod restarts with the fixed image, you should observe it is stable and there are no errors in the logs.

### CODE REVIEW

**Comment by israel-hdez** [comment]
@zdtsw Tagging you just FYI :-)

**Review by spolti — APPROVED** [review]
besides the changes required by Yuan, the changes looks good to me.

File: cmd/manager/main.go [hunk 1]
```diff
@@ -146,13 +147,27 @@ func main() {
 		log.Error(err, "unable to get ingress config.")
 		os.Exit(1)
 	}
-	if deployConfig.DefaultDeploymentMode == string(constants.Serverless) {
```
[the remark this answers is not in the document]
Comment (israel-hdez) [reply cmd/manager/main.go:149]: The upstream PR where all related changes are present should make it clear. I'll refer to it here to give clarity.

I'll take as an example the changes to the inference controller here: https://github.com/kserve/kserve/pull/3472/files#diff-f14605793ed521dabc5c4794584c2961365d2d8324ad41f009aa44b1513d50feR321.

Notice that an equivalent condition around `deployConfig.DefaultDeploymentMode` is also being removed in the controller.

So, this removal aligns the conditionals that were modified in the controllers (that were merged in #233).

File: cmd/manager/main.go [hunk 2]
```diff
@@ -146,13 +147,27 @@ func main() {
...
+
+	ksvcFound, ksvcCheckErr := utils.IsCrdAvailable(cfg, knservingv1.SchemeGroupVersion.String(), constants.KnativeServiceKind)
+	if ksvcCheckErr != nil {
+		log.Error(ksvcCheckErr, "error when checking if Knative Services are available")
```
[the remark this answers is not in the document]
Comment (israel-hdez) [reply cmd/manager/main.go]: Fixed

File: cmd/manager/main.go [hunk 3]
```diff
@@ -146,13 +147,27 @@ func main() {
...
+	if ksvcFound {
 		log.Info("Setting up Knative scheme")
 		if err := knservingv1.AddToScheme(mgr.GetScheme()); err != nil {
 			log.Error(err, "unable to add Knative APIs to scheme")
 			os.Exit(1)
 		}
-		if ingressConfig.DisableIstioVirtualHost == false {
+	}
+
+	if ingressConfig.DisableIstioVirtualHost == false {
```
[the remark this answers is not in the document]
Comment (israel-hdez) [reply cmd/manager/main.go]: Fixed

File: cmd/manager/main.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (terrytangyuan) [reply cmd/manager/main.go:149]: SGTM

**Comment by heyselbi** [comment]
Tested it in a OpenShift 4.14.13 cluster. Passed the test per instructions.

**Pull request merged**